### PR TITLE
Optimize display

### PR DESCRIPTION
**Pull request by drasmuss**
This addresses #276 (lag on firefox when dragging around large networks).  It's still a little bit choppy when dragging a big network around, but significantly improved from what it was like before.

The first improvement comes from only updating the minimap when it is being displayed.  Before we were updating it all the time, so we were effectively doing twice as many updates/redraws as necessary.

The most significant change to the code-base here is that I modified the minimap logic so that the main object takes care of keeping the minimap object up to date (rather than the old method, where the overall netgraph took care of updating both the main objects and the minimap objects).  This means that there is only one place where we need to check if the minimap is open, rather than having to remember to put that check in everywhere we update the object.

The second improvement comes from reducing the number of jquery calls we're making, as those are kind of expensive.  Basically this just involves storing the jquery result in a local variable when we query it (in response to something changing), and then everything else just references that local variable instead of redoing the jquery.  (note: I only did this in `netgraph.js` and `netgraph_item.js`; there are likely other areas of the code that could benefit from the same treatment).

### CODE REVIEW

**Comment by drasmuss** [comment]
Note: depends on #675

File: nengo_gui/static/components/netgraph_item.js [hunk 1]
```diff
@@ -548,15 +593,8 @@ Nengo.NetGraphItem.prototype.get_displayed_size = function() {
...
-        while (parent !== null) {
-            h_scale = h_scale * parent.size[0] * 2;
-            v_scale = v_scale * parent.size[1] * 2;
-            parent = parent.parent;
-        }
-
-        var w = this.size[0] * h_scale;
-        var h = this.size[1] * v_scale;
+    	var w = this.get_nested_width() * h_scale;
+    	var v = this.get_nested_height() * v_scale;
```
Comment (tcstewar) [inline nengo_gui/static/components/netgraph_item.js]: I think this should be ```var h``` rather than ```var v```....

Comment (drasmuss) [inline nengo_gui/static/components/netgraph_item.js]: Ah yeah, definitely.  The `v_scale` threw me off!

**Comment by drasmuss** [comment]
Hang on on merging/reviewing this one, I just noticed some weird things happen with recurrent connections that I need to look at.

**Comment by tcstewar** [comment]
> Hang on on merging/reviewing this one, I just noticed some weird things happen with recurrent connections that I need to look at.

Will do...  I was going through others and wanted to try this one out, and didn't notice that it wasn't marked as ready for review...  :)

**Comment by drasmuss** [comment]
Actually it was just solved by fixing the thing you already pointed out, so good to go!

**Comment by tcstewar** [comment]
Collapsing subnetworks seems to be broken.  To replicate, open up anything with subnetworks (e.g. ```tutorial/18-networks.py```), expand a network, and then collapse it again.  The subcomponents aren't removed and you get a javascript error of:

```
Uncaught NotFoundError: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
Nengo.NetGraphItem.remove @ netgraph_item.js:578
Nengo.NetGraphItem.remove @ netgraph_item.js:584
Nengo.NetGraphItem.collapse @ netgraph_item.js:505
(anonymous function) @ netgraph_item.js:326
```

**Comment by drasmuss** [comment]
Should be fixed now.  That actually revealed another inefficiency; we were adding all the minimap objects to the `network.children` list, so then when we went to redraw the children we'd redraw all the main objects and minimap objects, and then we'd redraw all the minimap objects again.  So it's even faster now!  Bugs only make me stronger :rage3:.

**Comment by tcstewar** [comment]
Well, I haven't found any other ways to break it, and all the changes look good to me, so I think this is ready for merging!

**Comment by tcstewar** [comment]
(and thanks for going through that code and profiling things and figuring out where some of the slowdowns were!)

**Comment by drasmuss** [comment]
I'd do that just for fun, optimization is like popping bubble wrap :heart_eyes:.

**Comment by tcstewar** [comment]
> I'd do that just for fun, optimization is like popping bubble wrap .

Wow did we hire the right person.  ;)

**Comment by Seanny123** [comment]
What did you use for profiling the code? Would it be worth documenting your methods somewhere?

**Comment by drasmuss** [comment]
Just the built-in Firefox profiler.  Chrome has a built-in profiler as
well, although it isn't quite as fully featured.

**Pull request merged**